### PR TITLE
pfblockerng: add a hysteresis margin to the log-retention trim (#1109)

### LINKED ISSUE #1109: Log retention: add a hysteresis margin so trims stop rewriting the whole log every tick (SSD write amplification)
## Summary

Add a single global setting, `pfb_log_trim_margin_pct`, that gives both log-retention caps (line-count **and** age) a hysteresis band: only rewrite a log file once it has drifted past its cap by the configured percentage, then trim it back to the exact cap. This eliminates the current behaviour where an opted-in (or default line-count) log gets its whole retained body rewritten on essentially **every** cron tick, which is needless SSD/CF write amplification.

Default `'0'` = today's behaviour preserved for anyone who doesn't opt in (with one strictly-better exception noted below).

## Motivation

`pfb_log_mgmt()` (`src/usr/local/pkg/pfblockerng/pfblockerng.inc:3006`) trims each log with a *tail-to-temp-then-cat-over-the-live-file* step that preserves the inode (#264/#280). The `cat temp > file` step **rewrites the entire retained body of the log**.

Two problems compound:

1. **The line-count cap (`log_max_<type>`, default `'20000'`) is always on** and, for any log with a numeric cap, the current code runs `tail -n N > temp; cat temp > file` **every idle tick regardless of whether anything changed** — there is no "already within cap" guard. `pfb_log_mgmt()` runs at the end of every *idle* tick (`pfblockerng_extra.inc:2722`; tick cadence `pfb_tick_interval`, default 15 min).
2. **The age cap (`log_max_days_<type>`, ADR-60) cannot be helped by a naive "skip if nothing expired" guard.** At steady state a log spans exactly the retention window, so its oldest line sits right at the cutoff; ~15 minutes later that head slice has expired, so `pfb_log_age_nolimit_pass_needed()` (`:2986`, added in #1052) reports work to do and the rewrite fires **every tick**. To evict ~15 min of head we rewrite the whole N-day body: for a 1-day cap at 15-min ticks that is writing ~96× the retained size per day just to shave slivers off the front.

We already explored evicting only the head in place (`memmove`-down + `ftruncate`, keeping the inode) — a possible future optimisation, but out of scope here because it changes the write path materially and has a torn-write window. **Frequency reduction (this issue) is the small, low-risk lever that directly addresses the write amplification.** The alternative of a calendar-boundary daily reset was rejected: it reintroduces the `log_rotate.last` marker file + wall-clock coupling that ADR-60 deliberately deleted (ADR-60 §1.1/§2.1) and brings back the clock-skew sawtooth of the retired ADR-30.

### Config

- New **global** registered scalar `pfb_log_trim_margin_pct` (not per-log — one knob applies to whichever cap is active on every log type).
  - A single global knob is only expressible as a **percentage**, because it must apply uniformly to both a *line* cap and a *days* cap. That is the deliberate choice; do not make it per-type or per-unit.
- Default `'0'`. Integer percent, stored as a string like the sibling keys.
- Parser: a new pure `pfb_log_trim_margin_pct(): int` mirroring `pfb_log_max_days()` (`pfblockerng.inc:2815`): `preg_match('/^[0-9]+$/', ...)` → `(int)`, clamped to a sane upper bound so `days*86400*(1+m/100)` cannot overflow; anything else (empty / non-numeric / negative / float / whitespace / array) → `0` (off). Place it beside `pfb_log_max_days()`.

### Semantics — high-water trigger, low-water cut (hysteresis)

Add **one** guard that decides whether to perform the existing rewrite at all. The rewrite itself (the `tail -n N` + age-trim + `cat`) is **unchanged** and still cuts to the exact cap (low-water).

```
rewrite this tick IFF:
      (line cap active AND line_count(file) > log_max      * (1 + margin/100))
   OR (age  cap active AND oldest_line(file) older than  log_max_days days * (1 + margin/100))
```

- When it fires, the existing path runs as today: `tail -n log_max` for the line cap, then `pfb_log_age_trim_temp()` cuts at exactly `now - log_max_days*86400` (low-water), then `cat` over the live file (inode preserved).
- The log therefore sawtooths between the cap (low-water) and `cap × (1 + margin/100)` (high-water). `margin=50` on a 1-day age cap ⇒ fires ~every 12 h instead of every tick; `margin=0` ⇒ fires whenever anything exceeds the cap (today's trigger).
- **This new guard subsumes the existing `pfb_log_age_nolimit_pass_needed()` skip (#1052)** — that becomes the `margin=0` age case. Generalise it (add an optional `int $margin_pct = 0`, apply the `(1 + margin/100)` multiplier to the cutoff; its "empty file / unparseable-or-legacy first line ⇒ TRUE (run the pass)" fallthrough is preserved). Rename to something cap-neutral (e.g. `pfb_log_age_trim_needed`) and update its callers + test, or keep the name with a param and a note — implementer's call, justified in the handoff.
- Add a small `pfb_log_line_count(string $path): int` helper (`wc -l`, or PHP), used only for the line high-water probe. On probe failure return a large value so a needed trim is never skipped (mirror the `#1052` "on failure ⇒ run the pass" safety direction). Note the ceiling: this reads the whole file each tick to decide — a **read**, not a write; SSD wear is dominated by writes, so this is acceptable and should carry a one-line `// ponytail:`-style comment naming the read-cost ceiling.

### The `margin=0` nuance (must be called out in the handoff + ADR)

With `margin=0`, the line-count high-water equals the cap, so the guard **skips the rewrite when the file is already within the cap** — whereas today the numeric-`log_max` path rewrites unconditionally. The resulting **log content is byte-identical** to today (a file already ≤ N lines with no expired head produces an identical `tail`/`cat`); the only change is that a redundant no-op full-file rewrite (and its mtime bump) is elided. This is intended — it is exactly the default-on line-count thrash the issue targets — but it is a real I/O-behaviour change from "rewrite every tick" to "rewrite only when out of band", so it must be documented, not smuggled in. Nothing is known to key on a log file's mtime (ADR-42 content-hashing is for feeds, not logs); confirm during implementation.

### Integration points (minimal diff — the write path stays byte-for-byte)

`pfb_log_mgmt()` has two structurally-parallel branches that both must get the guard:

- chroot Python logs (`dnslog`/`dnsreplylog`/`unilog`), `pfblockerng.inc:3029`–`:3076`
- plain logs, `:3078`–`:3104`

Read the global once at the top of `pfb_log_mgmt()` (after `pfb_global()`), the same way `log_max_*`/`log_max_days_*` are read from `$pfb['config'][...]`. Replace each branch's existing `#1052` nolimit-skip with the unified `rewrite this tick IFF ...` guard; leave the `tail`/`copy`/age-trim/`cat`, the per-exec exit-code gating, the short-write guard (#1053), and the `chown/chgrp unbound` (#264) exactly as they are.

### UI (`src/usr/local/www/pfblockerng/pfblockerng_general.php`, Log Settings section)

- Add one global numeric `Form_Input` for `pfb_log_trim_margin_pct` (min `0`, unit label "%"), placed at the **section level** (not inside the per-type `foreach` at `:412`) — e.g. just before the per-type table or right after the intro `Form_StaticText` at `:393`. Extend that intro help text to explain the margin (one to two sentences, matching neighbouring help-text style/length).
- Read via `PfbConfig::read('pfb_log_trim_margin_pct')` alongside the `log_max_days_*` reads (`:86`).
- Validate on POST like the `log_max_days_*` loop (`:159`–`:164`: `ctype_digit` else `'0'`), and persist into `$pfb['gconfig'][...]` before the `writeSection()` call (mirror `:214`–`:216`).

### Config-gateway lockstep (ADR-29 — all four in one change)

1. Registry entry in `pfb_cfg_registry()` (`src/usr/local/pkg/pfblockerng/pfblockerng_extra.inc`) — add `pfb_log_trim_margin_pct` to the global `$registry = $registry + [ ... ]` block near `pfb_software_check` (`:627`), `section => $gen`, `default => '0'`, adapters `NULL`, `since` matching the co-located ADR-60 keys' value (bump to the current dev PORTVERSION if it has advanced — read it from the siblings, do not invent).
2. Sniff path in `tests/phpcs/PfBlockerNG/Sniffs/Config/RequireConfigGatewaySniff.php` `$registeredPaths` (`:61`), alongside the `log_max_days_*` and `pfb_tick_interval` entries.
3. Round-trip coverage in `tests/php/CfgGatewayTest.php`.
4. UI reads/writes go through `PfbConfig` (above).

## Coverage matrix (enumerate from source, not memory)

| Axis | Rows |
| --- | --- |
| Parser `pfb_log_trim_margin_pct()` | `''`, `'0'`, valid `'50'`, non-numeric `'abc'`, negative `'-5'`, float `'1.5'`, leading/trailing whitespace, array input, oversized (clamp) |
| line cap × age cap × margin | (numeric line, age off, m=0) ⇒ rewrite only when lines>cap, **content-identical to today, redundant rewrite elided**; (numeric line, age off, m>0, under/over high-water); (nolimit, age set, m=0) ⇒ unchanged from #1052; (nolimit, age set, m>0, under/over); (numeric line, age set, m=0) ⇒ unchanged; (numeric line, age set, m>0) ⇒ fires iff **either** cap over its high-water, both cut to low-water; (nolimit, age off) ⇒ `continue` unchanged |
| Both `pfb_log_mgmt()` branches | chroot Python-log branch **and** plain-log branch each carry the guard |
| Contracts preserved | inode + `chown/chgrp unbound` preserved on every fire path (#264); failed `tail`/`copy`/`cat`/`wc` never blanks the live log (exit-code gating, #1061); short-write guard intact (#1053) |
| Age high-water probe | oldest-line parse reuses `pfb_log_age_field_mode()`/`pfb_log_line_timestamp()`; empty file ⇒ no fire; unparseable/legacy first line ⇒ fire (matches #1052) |

## Hostile-input rows

- `pfb_log_trim_margin_pct` set to `''` / `'abc'` / `'-5'` / `'1.5'` / `'  50 '` / an array ⇒ treated as `0` (off).
- Oversized numeric margin (e.g. `'999999999'`) ⇒ clamped; `days*86400*(1+m/100)` must not overflow PHP int.
- Line-count probe on a file with **no trailing newline** ⇒ off-by-one undercount is harmless at a `cap×(1+margin)` threshold; do not add complexity for it, but note it.
- Age high-water probe on an empty file and on a file whose first line is a legacy/pre-ADR-60 (unparseable) timestamp ⇒ preserve #1052's "run the pass" fallthrough.

## Tests

- **Red→green** for the behaviour change: seed a log at, say, `1.2×` its cap with `margin=50` (high-water `1.5×`) and assert **no rewrite occurs** (content + inode + mtime unchanged) — this **fails against current `pfb_log_mgmt()`** (which always rewrites) and passes after. Then seed at `1.6×` and assert it trims back to the cap.
- **Oracle (behaviour-preserving):** `margin=0` produces byte-identical **log content** to the pre-change trim across the matrix rows (the redundant-rewrite elision is asserted separately as the intended I/O change, e.g. mtime unchanged when already within cap).
- Mirror the existing suites: `tests/php/LogMaxDaysTest.php` (parser), `tests/php/LogMgmtAgeCutoffTest.php` + `tests/php/LogAgeCutoffTest.php` (trigger + cut), `tests/php/CfgGatewayTest.php` (registry round-trip), the `RequireConfigGateway` sniff test.
- **Tier-A `ui_render`** (mandatory for the `www/` change): extend `tests/smoke/ui/test_render_smoke.py` to assert the new field renders on the Log Settings page and the page is PHP-error-free.
- Optional live-VM smoke: extend `tests/smoke/test_log_age_retention.py` to prove a mid-window tick does **not** rewrite while a past-high-water tick does (inode preserved).

## Gates (per CLAUDE.md canonical table)

- PHP: `php -l` per touched file · `vendor/bin/phpunit` · `vendor/bin/phpstan` · `vendor/bin/phpcs --standard=phpcs.xml.dist src/`
- `www/`: Tier-A `ui_render` coverage exists for the change
- No Python touched, so pytest is unaffected (unless the smoke helper changes).

## ADR-60 §8 amendment (same change)

This extends ADR-60's retention mechanism, so amend `.ADRs/ADR_60_Age_Based_Log_Retention/ADR.md` §8 in the same PR: add an amendment entry describing the `pfb_log_trim_margin_pct` hysteresis (high-water trigger / low-water cut), the `margin=0` redundant-rewrite-elision nuance, that it applies to both caps, and the `wc -l` read-cost ceiling. (CLAUDE.md: a change extending Accepted ADR text amends the ADR in the same change.)

## Definition of done

- `pfb_log_trim_margin_pct` registered (registry + `since` + sniff `$registeredPaths` + round-trip test), default `'0'`.
- `pfb_log_mgmt()` both branches gated by the unified high-water trigger; existing inode/exit-code/short-write/ownership contracts untouched.
- UI field + help text; read/validate/save through `PfbConfig`.
- Full coverage + hostile-input matrix; red→green for the behaviour change; `margin=0` content-oracle green.
- All gates green; ADR-60 §8 amended.

## Non-goals

- In-place head eviction (`memmove`/`ftruncate`/`FALLOC_FL_COLLAPSE_RANGE` — the latter is Linux-only and unavailable on FreeBSD anyway). Possible future optimisation; not this issue.
- Any calendar-boundary / marker-file scheme (explicitly rejected — reintroduces what ADR-60 removed).
- Per-log-type margins (a single global percentage was the agreed scope).

## Key references

- `pfb_log_mgmt()` — `src/usr/local/pkg/pfblockerng/pfblockerng.inc:3006`; branches `:3029`–`:3076` (chroot Python logs) and `:3078`–`:3104` (plain logs).
- `pfb_log_max_days()` parser to mirror — `:2815`; `pfb_log_max_lines()` — `:2802`.
- `pfb_log_age_nolimit_pass_needed()` (#1052, to generalise) — `:2986`; `pfb_log_age_trim_temp()` — `:2957`; `pfb_log_age_field_mode()` — `:2828`; `pfb_log_line_timestamp()` — `:2848`.
- Tick call site + idle-tick gate — `src/usr/local/pkg/pfblockerng/pfblockerng_extra.inc:2722`; `pfb_tick_interval` default `'15'` — `:700`.
- Registry — `pfb_cfg_registry()` in `pfblockerng_extra.inc`; `log_max_days_*` loop `:614`–`:622`; global block near `pfb_software_check` `:627`.
- Sniff `$registeredPaths` — `tests/phpcs/PfBlockerNG/Sniffs/Config/RequireConfigGatewaySniff.php:61` (log_max_days paths `:83`–`:93`).
- UI — `src/usr/local/www/pfblockerng/pfblockerng_general.php`: help `:393`, reads `:86`, POST validate `:159`, save `:214`, per-type render `:412`–`:427`.
- ADR — `.ADRs/ADR_60_Age_Based_Log_Retention/ADR.md` (§1.1/§2.1 rationale, §8 amendments).

### PR DESCRIPTION
**Pull request by andrebrait**
Fixes #1109

## What

`pfb_log_mgmt()` trims each log with a *tail-to-temp-then-cat-over-the-live-file* step that preserves the inode (#264/#280) — and that `cat temp > file` rewrites the **entire retained body** of the log. Today it fires on essentially every idle tick:

- The line-count cap (`log_max_<type>`, default `20000`) is always on and had **no "already within cap" guard**, so a numeric cap ran `tail -n N > temp; cat temp > file` every tick regardless of whether anything changed.
- The age cap can't be helped by a naive "skip if nothing expired" check: at steady state the oldest line sits right at the cutoff, so ~15 minutes later `pfb_log_age_nolimit_pass_needed()` (#1052) reports work and the rewrite fires again. Evicting ~15 min of head means rewriting the whole N-day body — for a 1-day cap at 15-min ticks, roughly 96× the retained size written per day to shave slivers off the front.

This adds a single global knob, `pfb_log_trim_margin_pct` (default `'0'`), giving **both** caps a hysteresis band:

```text
rewrite this tick IFF:
      (line cap active AND line_count(file) >  log_max      * (1 + margin/100))
   OR (age  cap active AND oldest_line(file) older than log_max_days days * (1 + margin/100))
```

High-water **trigger**, low-water **cut**: when it fires, the existing write path is unchanged and still cuts to the *exact* cap. The log sawtooths between `cap` and `cap × (1 + margin/100)` — `margin=50` on a 1-day age cap fires about every 12 h instead of every tick.

## The `margin=0` nuance (intentional, not smuggled in)

With the default `margin=0` the line-count high-water equals the cap, so the guard **skips the rewrite when the file is already within the cap** — whereas the old code rewrote it unconditionally. The resulting **log content is byte-identical** (a file already ≤ N lines with no expired head produces an identical `tail`/`cat`); only the redundant no-op full-file rewrite and its mtime bump are elided. That is exactly the default-on line-count thrash this targets. Nothing keys on a log file's mtime (ADR-42 content-hashing covers feeds, not logs).

## How

- New pure parser `pfb_log_trim_margin_pct()`, mirroring `pfb_log_max_days()`'s shape (clamped to 1000).
- New `pfb_log_line_count()` — pure PHP, `PHP_INT_MAX` on any probe failure so a needed trim is never skipped (same fail-open direction as #1052).
- `pfb_log_age_nolimit_pass_needed()` → **`pfb_log_age_trim_needed()`** (the probe is no longer `nolimit`-only) with an `int $margin_pct = 0` parameter. Its empty-file / unparseable-first-line / unopenable fallthroughs are preserved verbatim.
- **`pfb_log_trim_needed()` is now the single guard for both `pfb_log_mgmt()` branches**, replacing the two duplicated #1052 skip lines. It subsumes #1052's skip as the `margin=0` age case.
- The line high-water uses float math deliberately: `pfb_log_max_lines()` applies no clamp, so `$logmax` can reach `PHP_INT_MAX` from a typo'd config, and `intdiv()` would receive a float (PHP promotes integer overflow to float) and throw a `TypeError` in the cron tick.
- Registry + `RequireConfigGateway` sniff + `CfgGatewayTest` round-trip, per ADR-29.
- UI: one section-level field in Log Settings, read/validated/persisted through `PfbConfig`, with the intro help text extended.
- Everything after the guard — `tail`/`copy`, the age trim, the `cat` over the live file, per-exec exit-code gating (#1061), the short-write guard (#1053), and `chown`/`chgrp unbound` (#264) — is byte-for-byte unchanged.

## Tests

Red→green, test-first: reproduction tests authored and executed RED against the untouched code (a 1.2×-cap file with `margin=50` gets truncated today, so content is the oracle; the elision row uses a forced-old mtime), frozen by hash, then GREEN with zero edits.

- `LogTrimMarginPctTest` (13 hostile/valid parser rows), `LogLineCountTest` (4), `LogAgeTrimNeededTest` (7), `LogTrimNeededTest` (14, incl. the absurd-`log_max` `TypeError` trap), `LogMgmtAgeCutoffTest` (+9 integration rows driving real `pfb_log_mgmt()` on **both** branches), `CfgGatewayTest` (+3).
- Tier-A `ui_render` and Tier-B `ui_e2e` executed on a live pfSense VM: the field renders, the save round-trips, every rejected shape plus an array-valued POST (the #1143 class) coerces to `0` without a `TypeError`, and an out-of-range value is clamped **on save** — the UI POST is canonicalized through the same `pfb_log_trim_margin_pct()` parser the backend reads with, so the stored value is always the effective one.

## ADR

Amends ADR-60 §8 in the same change (the hysteresis, why a percentage, the `margin=0` elision, the rename, the read-cost trade, and the rejected alternatives — in-place head eviction and any calendar-boundary/marker-file scheme).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017Vv1F4fW9zu39cSJ69Jn7f

## Summary by CodeRabbit

* **New Features**
  * Added a configurable log-trimming margin percentage, from 0–1000%, in General settings.
  * The margin applies to both line-count and age-based log limits.
  * Logs are trimmed back to the configured limits once trimming is triggered.

* **Bug Fixes**
  * Avoids unnecessary log rewrites when logs are already within configured limits.
  * Improved handling of invalid or unexpected setting values without errors.

* **Tests**
  * Added coverage for configuration validation, log trimming behavior, edge cases, and General page rendering.

### CODE REVIEW

**Comment by andrebrait** [comment]
@coderabbitai review

**Comment by andrebrait** [comment]
## Review audit — PR #1254

**Reviews received**

| Reviewer | Model / effort | Outcome |
| --- | --- | --- |
| Claude adversarial (`review-single`) — full diff | Sonnet, `xhigh` | 1 blocking, 1 nitpick |
| Claude adversarial — re-review of fix delta `6084b77a` | Sonnet, `xhigh` | 0 blocking, 3 nitpicks |
| Claude adversarial — re-review of fix delta `0760d15f` | Sonnet, `xhigh` | 0 blocking, 2 nitpicks |
| Claude adversarial — re-review of the CodeRabbit-fix delta | Sonnet, `xhigh` | 0 blocking, 1 nitpick |
| CodeRabbit | — | 2 findings (arrived late, caught by the pre-merge sweep) |
| GitHub Copilot | — | Reviewed 14/14 files, generated no comments |
| Snyk | — | No status on the head SHA — not active on this PR |

Model choice: `sonnet` at `xhigh` per the repo's reviewer contract (Fable, normally preferred for a diff this size, is temporarily unavailable). PR size at first review: 14 files, ~1141 insertions.

**Findings ledger**

1. **BLOCKING — `pfb_log_line_count()` undercounted a log left mid-line, silently skipping a needed trim.** It counted `\n` bytes, but the guard it feeds gates a `tail -n N` rewrite, and `tail` counts an unterminated trailing chunk as a line. A log whose last byte is not `\n` — the state `pfb_logger('.', 1)`'s progress fragments leave (15 call sites; `pfb_logger_target_starts_at_bol()` exists precisely to detect it) — read as "at cap" when it was really one line over, so the trim never fired. This falsified the change's central claim that `margin=0` leaves log content byte-identical. Reproduced independently before fixing. → **fixed@`6084b77a`**, refined@`0760d15f`, test-first red→green (reproduction tests authored and run RED against unfixed code, frozen by hash, GREEN with zero edits).
2. **nitpick — `$margin_pct` params trusted unclamped from the caller.** → **skipped**: `pfb_log_mgmt()` is the sole caller and always passes `pfb_log_trim_margin_pct()`'s clamped output; PHPStan enforces the `int` type. A second clamp inside the guard would be a third copy of a contract that already lives in the parser and its tests — the drift CLAUDE.md warns about. The reviewer found no reachable path.
3. **nitpick — src comment restated the ADR verbatim instead of pointing at it.** → **fixed@`0760d15f`**.
4. **nitpick — the bol-helper fails open to `TRUE`, which in the new use means `+0`: the undercount direction the fix exists to remove.** → **fixed@`0760d15f`** by tracking the final byte of the streaming read already being performed. Drops a second `fopen()`, closes the fail-open gap, and keeps memory bounded at one 1 MiB chunk.
5. **nitpick — no boundary test for a single unterminated line with zero newline bytes.** → **fixed@`0760d15f`**.
6. **nitpick — commit message said "one-line pointer"; the comment is two lines.** → **skipped**: the reviewer confirmed the comment itself is within budget; message wording only.
7. **nitpick — `LogLineCountTest` docblock still described a `\n` byte count.** → **fixed@`56567fba`** (stale text this PR itself introduced).
8. **CodeRabbit (Minor) — out-of-range margin persisted raw.** `ctype_digit('999999999')` is true, so config.xml stored `999999999` while the runtime parser clamped to `1000` — the field re-rendered a number the system never uses. → **fixed@`8dae53db`**: the POST is now canonicalized through the same `pfb_log_trim_margin_pct()` parser the backend reads with, so the stored value is the effective one. Net deletion (the hand-rolled `ctype_digit` guard goes away — one definition of the contract, not two). Test-first: Tier-B assertion committed alone and run **RED on a live VM**, then GREEN.
9. **CodeRabbit (Trivial) — gateway round-trip only covered `PfbConfig::write()`, not the `writeSection()` path the UI actually saves through.** Tier-B covers the real save but is **schedule-only, not the PR gate**, so nothing PR-gating covered it. → **fixed@`8dae53db`** (`testPfbLogTrimMarginPctSurvivesSectionWrite`; the reviewer confirmed it is non-vacuous by injecting a key-drop regression and watching it go red).
10. **nitpick — stale `:243` line pointer** (my own canonicalization fix shrank the file and moved the call to `:240`). → **fixed@`1ad17140`** by naming the symbols instead of the lines.

**Gates on the final head:** `php -l` · PHPUnit **3176 tests, 0 failures** · PHPStan `No errors` · PHPCS clean · markdownlint 0 errors · comment-narration clean. Live pfSense VM: Tier-A `ui_render` **4 passed**, Tier-B `ui_e2e` **8 passed**.

**Process note, recorded rather than buried:** the blocking finding (#1) was missed by both the delegated step's independent verifier and my own gate, and was caught only by the adversarial review. CodeRabbit's two findings arrived *after* it had twice reported a quota limit (17 min, then 47 min) and would have been dropped under the 5-minute rule — the pre-merge catch-all sweep is what surfaced them.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: src/usr/local/www/pfblockerng/pfblockerng_general.php [hunk 1]
```diff
@@ -168,6 +172,14 @@
...
 		}
 
+		// issue #1109: validate the log-trim hysteresis margin the same way (non-negative
+		// integer string; the backend parser independently clamps it to 0-1000).
+		$margin_v = $_POST['pfb_log_trim_margin_pct'] ?? '0';
+		if (is_array($margin_v) || !ctype_digit((string) $margin_v)) {
+			$margin_v = '0';
+		}
+		$_POST['pfb_log_trim_margin_pct'] = (string) $margin_v;
+
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/www/pfblockerng/pfblockerng_general.php]: Applied in `8dae53db`. You are right that the runtime clamp on read was not enough: `config.xml` stored `999999999` while the effective margin was `1000`, and the field re-rendered the raw value — showing a number the system never uses.

Rather than add a bound check alongside the `ctype_digit()` guard, the POST is now canonicalized through the *same* parser the backend reads with:

```php
$_POST["pfb_log_trim_margin_pct"] = (string) pfb_log_trim_margin_pct($_POST["pfb_log_trim_margin_pct"] ?? "0");
```

That replaces the hand-rolled guard entirely (net deletion), so there is one definition of the contract instead of two that can drift, and the stored value is always the effective one. The parser's `is_scalar()` check short-circuits before any string cast, so the array-valued POST case (#1143 TypeError class) stays closed.

Pinned test-first: the Tier-B assertion (`999999999` → `1000`) was committed alone and run RED on a live pfSense VM against the unchanged save path, then GREEN after the fix.

**Comment by andrebrait** [comment]
On the nitpick about `tests/php/CfgGatewayTest.php` (section-level write path) — applied in `8dae53db`.

The point holds for a reason worth stating: the Tier-B `ui_e2e` flow does exercise the real UI save (a CSRF-carrying POST, then reading `config.xml` back off the box), but Tier B is **schedule/dispatch-only and is not the PR gate** in this repo. So the only PR-gating coverage of the `writeSection()` path was... none. A key silently dropped on the section write would have shipped green.

`testPfbLogTrimMarginPctSurvivesSectionWrite` now exercises `readSection()`/`writeSection()` with the exact section string the page saves with, asserting the before-state first. It was checked for vacuity by injecting a key-drop regression into `writeSection()` — the test goes red on it, which is the regression class it exists to guard.
